hs_explore/utils.py

```diff
@@ -0,0 +1,248 @@
+from django.contrib.auth.models import User
+from hs_core.models import BaseResource
+from hs_tracking.models import Variable
+from django.db.models import Max
+from collections import defaultdict
+from hs_core.hydroshare.utils import user_from_id, get_resource_by_shortkey
+from hs_explore.models import RecommendedResource, UserPreferences, Status
+import string
+from nltk.stem import WordNetLemmatizer
+from hs_explore.models import LDAWord
+
+
+def user_resource_matrix(fromdate, todate):
+    """ return a list of (username, resource_id, last_access_timestamp), which
+        shows the latest time that a user accessed a resource during the given
+        period.
+        :param fromdate (date type), the start date of the time period
+        :param todate (date, type), the end date of the time period
+    """
+    users = User.objects.filter(
+        visitor__session__variable__timestamp__gte=fromdate,
+        visitor__session__variable__timestamp__lte=todate)\
+            .distinct()
+    events = []
+    for u in users:
+        resources = BaseResource.objects.filter(
+                variable__timestamp__gte=fromdate,
+                variable__timestamp__lte=todate,
+                variable__session__visitor__user=u)
+        for r in resources:
+            latest = Variable.objects.filter(
+                    resource=r,
+                    session__visitor__user=u,
+                    timestamp__gte=fromdate,
+                    timestamp__lte=todate)\
+                    .aggregate(latest=Max('timestamp'))
+
+            events.append((u.username, r.short_id, latest['latest']))
```

Comment:
These should be optimized following the same ideas I outlined elsewhere in the review.

Reply:
I think I need some help on how to optimize this. Can you offer a help with it？ Thanks!

Reply:
This is a tricky one because of the aggregate.  But you can group by with values and then use annotate to "aggregate" for each group.

```
def user_resource_matrix(fromdate, todate):
	user_ids = User.objects.filter(
	visitor__session__variable__timestamp__gte=fromdate,
	visitor__session__variable__timestamp__lte=todate)\
		.distinct().values_list("pk", flat=True)
	
	res_ids = BaseResource.objects.filter(
		variable__timestamp__gte=fromdate,
		variable__timestamp__lte=todate,
		variable__session__visitor__user__id__in=user_ids).values_list("pk", flat=True)
		
	latest = Variable.objects.filter(
				resource__id__in=res_ids,
				session__visitor__user__id__in=user_ids,
				timestamp__gte=fromdate,
				timestamp__lte=todate)
	grouped = latest.values('session__visitor__user__username', 'resource__short_id')
	grouped_with_latest = grouped.annotate(latest=Max('timestamp'))
	return grouped_with_latest.values_list('session__visitor__user__username', 'resource__short_id', 'latest')
````

Reply:
Agree. This avoids the loop for using "aggregate".